RDMA/bnxt_re: Trim the QP mtu if it is higher than L2 MTU

bgooty opened the pull request:
Gen P5 chip allow sending roce packets bigger than L2 MTU.
But at the responder the packets are dropped if mismatch is
detected in the incoming packet size and the interface MTU.
If this is attempted by a malicious or badly written code
the chip drops incoming packets and corresponding rx error
counter is incremented. Since there is no response from
the responder QP the requester QP moves to ACK-Timeout
state.

Fixing the QP mtu setting to trim the qp-mtu to the lower
bound of L2 MTU during modify-qp from init to rtr.

Temporarily driver would fail the modify-request with mis-
matching MTU due to unavilability of a method which would
tell user qp about trimmed MTU.

Moved the mtu initialization code to separate inline
functions for ease of flow.

Signed-off-by: Bharat Gooty <bharat.gooty@broadcom.com>

bharat-gooty committed:
RDMA/bnxt_re: Trim the QP mtu if it is higher than L2 MTU

Gen P5 chip allow sending roce packets bigger than L2 MTU.
But at the responder the packets are dropped if mismatch is
detected in the incoming packet size and the interface MTU.
If this is attempted by a malicious or badly written code
the chip drops incoming packets and corresponding rx error
counter is incremented. Since there is no response from
the responder QP the requester QP moves to ACK-Timeout
state.

Fixing the QP mtu setting to trim the qp-mtu to the lower
bound of L2 MTU during modify-qp from init to rtr.

Temporarily driver would fail the modify-request with mis-
matching MTU due to unavilability of a method which would
tell user qp about trimmed MTU.

Moved the mtu initialization code to separate inline
functions for ease of flow.

Signed-off-by: Bharat Gooty <bharat.gooty@broadcom.com>

nareshkumarpbs commented on drivers/infiniband/hw/bnxt_re/ib_verbs.c:
```diff
@@ -1821,17 +1821,21 @@ static int bnxt_re_modify_shadow_qp(struct bnxt_re_dev *rdev,
 	if (qp_attr_mask & ~IB_QP_ATTR_STANDARD_BITS)
 		return -EOPNOTSUPP;
 
+	qp = to_bnxt_re(ib_qp, struct bnxt_re_qp, ib_qp);
```
upstream does not use to_bnxt_re

nareshkumarpbs commented on drivers/infiniband/hw/bnxt_re/ib_verbs.c:
```diff
@@ -1943,18 +1907,19 @@ int bnxt_re_modify_qp(struct ib_qp *ib_qp, struct ib_qp_attr *qp_attr,
+		rc = bnxt_re_init_qpmtu(qp, rdev->netdev->mtu, qp_attr_mask,
+					qp_attr);
+		if (rc) {
+			dev_err(rdev_to_dev(rdev), "qp %#x invalid mtu",
```
dev_xxx is not used in upstream, instead use ibdev_xxx

bgooty replied:
Yeah. Thanks . got it

bharat-gooty committed:
RDMA/bnxt_re: Update HWRM to fix cmdq_query_func response

Update HWRM ROCE to fix cmdq_query_func response to handle
var-size WQE and backward compatible modes.

Signed-off-by: Bharat Gooty <bharat.gooty@broadcom.com>
